### ISSUE TITLE
New Version Completely Changed My Redirect URI

### ISSUE THREAD

**Issue by rkuhlke** [fedora-infra/flask-oidc#21]
We use the OVERWRITE_REDIRECT_URI and the OIDC_CALLBACK_ROUTE parameters and 2.0.2 redirected our app to use a completely different callback route than what is configured. We had to migrate our app back down to version 1.4.0 to fix this issue. Please address

**Comment by willthetanman** [comment]
@abompard We are also experiencing this issue 

> The callback route (aka "redirect URL") is not configurable with OIDC_CALLBACK_ROUTE anymore. It is always /authorize

from: [https://github.com/fedora-infra/flask-oidc/blob/develop/docs/changelog.rst](https://github.com/fedora-infra/flask-oidc/blob/develop/docs/changelog.rst)


I'm not sure if this is true as the redirect_uri we typically use is `/oidc/callback`. Can you review this change as we are also stuck reverting back to `1.4.0` to migrate outage in authflow.

**Comment by abompard** [comment]
Hey folks!

Would it work for you to change (or add) the new `<prefix>/authorize` route to your redirect URLs in your OIDC provider?
Also, could you please describe the use case of `OVERWRITE_REDIRECT_URI`? Why you'd want it to be anything different from the built-in redirect uri.

Thanks. I'm basically rebuilding flask-oidc on top of authlib and I'd like to minimize the amount and complexity of features I'll need to maintain in the long run.

**Comment by willthetanman** [comment]
While we can add the route `<prefix>/authorize` to our redirect URIs and the auth flow will now work, however this seems to not be conformant with OIDC specifications where typically its `<prefix>/callback`. I can only assume many others are impacted by this change if this is the case.  

As for our specific use case for `OVERWRITE_REDIRECT_URI`, we use it to force redirection to `https`

**Comment by aswinkevin** [comment]
@abompard I'm using accept token decorator , not using the require login , so the accept token works like a charm but how to get the token info inside the function ? what is the function to get token info ? 

token info means the username , email and etc.

`@app.route('/welcome', methods=['GET'])`
`@oidc.accept_token(scopes=["profile"])`
`def welcome_user():`
    `<how to get token info>`
    `return jsonify(message=f"welcome to homepage")`

**Comment by abompard** [comment]
> While we can add the route `<prefix>/authorize` to our redirect URIs and the auth flow will now work, however this seems to not be conformant with OIDC specifications where typically its `<prefix>/callback`. I can only assume many others are impacted by this change if this is the case.

Actually there's no recommended URL pattern in the OIDC spec, or at least none that I know of. Would you have a reference to the part of the spec that recommends `<prefix>/callback`? In my understanding, any redirect URL is fine as long as it's allowed on the identity provider side.

> As for our specific use case for `OVERWRITE_REDIRECT_URI`, we use it to force redirection to `https`

Ah I see, so you have visitors accessing the site in plaintext HTTP, and when they click "login" you want them to come back on HTTPs? Did I understand correctly? If so, do you need to maintain the plaintext access or could you redirect all plaintext connections to HTTPS? Since it has to be working for the callback URL anyway. Or did I miss something?

**Comment by abompard** [comment]
> @abompard I'm using accept token decorator , not using the require login , so the accept token works like a charm but how to get the token info inside the function ? what is the function to get token info ?
> token info means the username , email and etc.

Hey! This is currently not as simple as I'd like to. What you can do:

```python
from authlib.integrations.flask_oauth2 import current_token
profile = g._oidc_auth.userinfo(token=current_token)
```

This `profile` variable should contain the user info.

I've updated [the docs](https://github.com/fedora-infra/flask-oidc/blob/develop/docs/index.rst#resource-server).

**Comment by willthetanman** [comment]
> Actually there's no recommended URL pattern in the OIDC spec, or at least none that I know of. Would you have a reference to the part of the spec that recommends <prefix>/callback? In my understanding, any redirect URL is fine as long as it's allowed on the identity provider side.

I don't think its in the spec, but most documentation I've seen will recommend to use `<prefix>/callback` eg. https://auth0.com/docs/authenticate/identity-providers/enterprise-identity-providers/oidc. If thats the case why not allow flask-oidc to also accept this as opposed to having to add additional route `<prefix>/authorize`, as updating the module will break the flow of many existing configurations. 

> Ah I see, so you have visitors accessing the site in plaintext HTTP, and when they click "login" you want them to come back on HTTPs? Did I understand correctly? If so, do you need to maintain the plaintext access or could you redirect all plaintext connections to HTTPS? Since it has to be working for the callback URL anyway. Or did I miss something?

Yes this is correct in how we are using `OVERWRITE_REDIRECT_URI`

**Comment by april** [comment]
Just lost half a day tracking this issue down. Eep.

I am very surprised that an updated version would completely drop functionality like this, without at least a year or two of throwing warnings that it would eventually break. Especially given that the public documentation still indicates that `OIDC_CALLBACK_ROUTE` is fully supported.

Okta, probably the most popular non-platform IdP, uses `/authorization-code/callback` all over the place. Which is what I had been using for years until a recent deployment pulling in version 2.x of Flask-OIDC broke piles of integrations.

I'd also note that 2.x expanded the default `OIDC_SCOPES` as well, so now it is requesting `profile` even though it previously only pulled `openid` and `email`.

**Comment by abompard** [comment]
> Just lost half a day tracking this issue down. Eep.
> I am very surprised that an updated version would completely drop functionality like this, without at least a year or two of throwing warnings that it would eventually break.

Sorry about that, really. I didn't realize there was so many users of this library, since it has been broken for a couple years (since the release of Itsdangerous 2.1).
In theory, `OIDC_CALLBACK_ROUTE` should still be supported with a deprecation warning, and there's a redirect in place to the authorize view. In practice I realize that this does not work since the login view is going to use the authorize view as a redirect_url, not the `OIDC_CALLBACK_ROUTE`. I'll try to fix that.

> Especially given that the public documentation still indicates that `OIDC_CALLBACK_ROUTE` is fully supported.

Yeah I didn't haven't got access to the public doc yet, the former maintainer is a very busy person.
 
> Okta, probably the most popular non-platform IdP, uses `/authorization-code/callback` all over the place. Which is what I had been using for years until a recent deployment pulling in version 2.x of Flask-OIDC broke piles of integrations.

Interesting, does Okta mandate this callback URL, or make it complex to change it?

> I'd also note that 2.x expanded the default `OIDC_SCOPES` as well, so now it is requesting `profile` even though it previously only pulled `openid` and `email`.

Good point, thanks.

**Comment by abompard** [comment]
I think all issues would be addressed by #26. Please have a look.

I did not restore `OVERWRITE_REDIRECT_URI` but all `redirect_uri`s will now be in HTTPS, as the spec actually requires (for web applications such as Flask).

**Comment by april** [comment]
> Sorry about that, really. I didn't realize there was so many users of this library, since it has been broken for a couple years (since the release of Itsdangerous 2.1).

300,000 downloads per month, congrats on owning a library. 😂

> Interesting, does Okta mandate this callback URL, or make it complex to change it?

Nope, Okta neither mandates it nor is it particularly complicated to change. The problem isn't really technical but organizational; at most medium to large size companies, the team that maintains the Okta configurations is not the team that is writing web services.

At those places (and my employer is one of them), changing the callback URLs often means opening a ticket, waiting for it to make it into the next sprint, and then getting it fixed a week or two later.

In any case, I really appreciate you taking the time to address these fixes and for adopting an abandoned library.

**Comment by abompard** [comment]
Yeah it makes sense, thanks for taking the time to explain.

**Comment by abompard** [comment]
This should be fixed by version 2.0.3, published a couple hours ago.

**Comment by aswinkevin** [comment]
> > @abompard I'm using accept token decorator , not using the require login , so the accept token works like a charm but how to get the token info inside the function ? what is the function to get token info ?
> > token info means the username , email and etc.
> 
> Hey! This is currently not as simple as I'd like to. What you can do:
> 
> ```python
> from authlib.integrations.flask_oauth2 import current_token
> profile = g._oidc_auth.userinfo(token=current_token)
> ```
> 
> This `profile` variable should contain the user info.
> 
> I've updated [the docs](https://github.com/fedora-infra/flask-oidc/blob/develop/docs/index.rst#resource-server).

it works thanks :)
